crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -201,13 +201,53 @@ impl SettlementEncoder {
                 )?
             }
             OrderClass::Limit => {
-                anyhow::bail!("limit orders are not supported by the SettlementEncoder");
+                // Solvers calculate with slightly adjusted amounts compared to this order but because
+                // limit orders are fill-or-kill we can simply use the total original `sell_amount`.
+                let executed_amount = order.data.sell_amount;
+                let buy_price = self.custom_price_for_limit_order(&order)?;
+
+                self.add_custom_price_trade(
+                    order,
+                    executed_amount,
+                    scaled_unsubsidized_fee,
+                    buy_price,
+                )?
             }
         };
         self.pre_interactions.extend(interactions.pre.into_iter());
         Ok(execution)
     }
 
+    /// Uses the uniform clearing prices to compute the individual buy token price to satisfy the
+    /// original limit order which was adjusted to account for the `surplus_fee` (see
+    /// `compute_synthetic_order_amounts_if_limit_order()`).
+    /// Returns an error if the UCP doesn't contain the `sell_token` or if under- or overflows
+    /// happen during the computation.
+    fn custom_price_for_limit_order(&self, order: &Order) -> Result<U256> {
+        // The order passed into this function is the original order signed by the user.
+        // But the solver actually computated a solution for an order with `sell_amount -= surplus_fee.
+        // That means this is the equation the solver tried to satisfy:
+        // uniform_sell_price * (sell_amount - surplus_fee) = uniform_buy_price * buy_amount
+        //
+        // But actually this is the equation the solution has to satisfy:
+        // uniform_sell_price * sell_amount = custom_buy_price * buy_amount
+        // That means we can solve for the custom_buy_price like this:
+        // custom_buy_price = (uniform_sell_price * sell_amount) / buy_amount
```

Comment:
Something feels off here since the the buy price is independent of the `uniform_buy_price`...

Reply:
Does this make more sense?
1) Use UCP buy and sell price to compute the actual `buy_amount` adjusted for the `fee_surplus` (it should be lower)
2) Use the existing formula but with the new reduced `buy_amount`
`custom_buy_price = (uniform_sell_price * sell_amount) / reduced_buy_amount`

Reply:
Yes, and I think you also need to account for order side (sell/buy). I.e., compute the buy_amount for sell orders, and sell_amount for buy orders.

Reply:
@harisang... do you mind double-checking our logic here 😅

Reply:
This is how I understand the problem after the last discussion:
```rust
    fn custom_price_for_limit_order(&self, order: &Order) -> Result<U256> {
        let uniform_buy_price = self
            .clearing_prices
            .get(&order.data.buy_token)
            .context("buy token price is missing")?;
        let uniform_sell_price = self
            .clearing_prices
            .get(&order.data.sell_token)
            .context("sell token price is missing")?;

        let (sell_amount, buy_amount) = match order.data.kind {
            // This means sell as much `sell_token` as needed to buy exactly the expected
            // `buy_amount`. Therefore we need to solve for `sell_amount`.
            OrderKind::Buy => {
                let sell_amount = order.data.buy_amount * uniform_buy_price / uniform_sell_price;
                // We have to sell slightly more `sell_token` to capture the `surplus_fee`
                let sell_amount_adjusted_for_fees = sell_amount + order.metadata.surplus_fee;
                (sell_amount_adjusted_for_fees, order.data.buy_amount)
            }
            // This means sell ALL the `sell_token` and get as much `buy_token` as possible.
            // Therefore we need to solve for `buy_amount`.
            OrderKind::Sell => {
                // Solver actually used this `sell_amount` to compute prices.
                let sell_amount = order.data.sell_amount - order.metadata.surplus_fee;
                let buy_amount = sell_amount * uniform_sell_price / uniform_buy_price;
                (order.data.sell_amount, buy_amount)
            }
        };

        let custom_buy_price = sell_amount * uniform_sell_price / buy_amount;
        Ok(custom_buy_price)
    }
```

I didn't use checked math operations to keep it succinct for this comment.
Also in the second to last line I still have to add `order.data.fee_amount` in case we want to use a non 0 fee at some point in the future, right?

Reply:
This looks correct to me now.

But... its price related so 🤷.

Reply:
Did some napkin math and the equations look correct to me ✅.